src/session.rs

```diff
@@ -233,3 +230,28 @@ impl Session {
         Ok(request.build_response(Some(remote_err)))
     }
 }
+
+/// Parse the consensus state from an incoming request
+// TODO(tarcieri): fix the upstream Amino parser to do this correctly for us
+fn parse_request<R>(request: &R) -> Result<(SignedMsgType, consensus::State), Error>
+where
+    R: TendermintRequest + Debug,
+{
+    let msg_type = match request.msg_type() {
+        Some(ty) => ty,
+        None => Err(err!(ProtocolError, "no message type for this request"))?,
+    };
+
+    let mut consensus_state = match request.consensus_state() {
+        Some(state) => state,
+        None => Err(err!(ProtocolError, "no consensus state in request"))?,
+    };
```

Comment:
Note that this relies on every message we sign including both of these, but I think that's the case anyway?